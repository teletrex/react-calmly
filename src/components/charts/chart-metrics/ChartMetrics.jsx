/* <LICENSE>
*
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
*
* </LICENSE>
*  */



import React, { useEffect, useState, useCallback } from 'react';
import PropTypes from 'prop-types';
import  settings  from '../../../settings';
import cloneDeep from 'lodash/cloneDeep';
import find from 'lodash/find';
import omit from 'lodash/omit';
import filter from 'lodash/filter';
import classNames from 'classnames';

import {Dropdown, MultiSelect} from '@carbon/react';  // was new-dropdown
import { useTranslation } from '../../../translation';

const { prefix } = settings;

const ChartMetrics = ({ items, onChange, limit, width, placeholder, label, helperText, size = 'sm'}) => {
  const { t } = useTranslation();
  const [metrics, setMetrics] = useState([]);
  const [selectedMetrics, setSelectedMetrics] = useState([]);
  const isMultiselect = limit > 1;

  const omitProperty = (data, prop) => omit(data, prop);
  const omitChecked = useCallback(data => data.map(i => omitProperty(i, 'checked')), []);

  const setItemDisabledStatus = useCallback(
    (data, selectedMetrics) => {
      const hitLimit = isMultiselect && selectedMetrics.length >= limit;
      return data.map(metric => ({
        ...metric,
        isDisabled: metric.isDisabled || (!find(selectedMetrics, metric) && hitLimit),
      }));
    },
    [isMultiselect, limit]
  );

  useEffect(() => {
    const itemsClone = omitChecked(cloneDeep(items || []));
    setMetrics(setItemDisabledStatus(itemsClone, selectedMetrics));
  }, [items, omitChecked, selectedMetrics, setItemDisabledStatus]);

  useEffect(() => {
    const nextSelectedMetrics = omitChecked(filter(items, 'checked'));
    setSelectedMetrics(setItemDisabledStatus(nextSelectedMetrics, nextSelectedMetrics));
  }, [items, omitChecked, setItemDisabledStatus]);

  const handleMetricSelect = e => {
    const hasOnChange = onChange && typeof onChange === 'function';
    if (isMultiselect) {
      const { selectedItems } = e;
      if (selectedItems.length <= limit) {
        if (hasOnChange) {
          onChange(selectedItems.map(item => omitProperty(item, 'isDisabled')));
        }
        setSelectedMetrics(selectedItems);
      }
    } else {
      const { selectedItem } = e;
      if (hasOnChange) {
        onChange(omitProperty(selectedItem, 'isDisabled'));
      }
      setSelectedMetrics([selectedItem]);
    }
  };

  const style = width
    ? {
        maxWidth: `${width}px`,
      }
    : null;

  return (
    <div
      className={classNames(`${prefix}--chart-metrics`, {
        [`${prefix}--chart-metrics-multiple`]: isMultiselect,
      })}
      data-testid="chart-metrics"
      style={style}
    >
      {isMultiselect &&
      <MultiSelect
        size={size}
        ariaLabel="Dropdown"
        helperText={
          helperText || t('Select up to {{limit}} metrics', { limit })
        }
        id="chart_metrics"
        items={metrics}
        itemToString={item => (item ? item.name : '')}
        labelText={label || t('Metrics')}
        onChange={handleMetricSelect}
        placeholder={placeholder || t('selected')}
        selectedItems={isMultiselect ? selectedMetrics : selectedMetrics[0]}
      />

      }
      { !isMultiselect &&
      <Dropdown
        size={size}
        ariaLabel="Dropdown"
        helperText={ helperText }
        id="chart_metrics"
        items={metrics}
        itemToString={item => (item ? item.name : '')}
        labelText={label || t('Metrics')}
        onChange={handleMetricSelect}
        placeholder={placeholder || t('selected')}
        selectedItem={isMultiselect ? selectedMetrics : selectedMetrics[0]}
      />
      }
    </div>
  );
};

export const chartMetricsPropTypes = {
  helperText: PropTypes.string,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      checked: PropTypes.bool.isRequired,
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
    })
  ),
  label: PropTypes.string,
  limit: PropTypes.number,
  onChange: PropTypes.func,
  placeholder: PropTypes.string,
  width: PropTypes.number,
};

export const chartMetricsDefaultPropTypes = {
  helperText: '',
  items: [],
  label: '',
  limit: 2,
  onChange: undefined,
  placeholder: '',
  width: undefined,
};

ChartMetrics.propTypes = chartMetricsPropTypes;
ChartMetrics.defaultProps = chartMetricsDefaultPropTypes;

export default ChartMetrics;
