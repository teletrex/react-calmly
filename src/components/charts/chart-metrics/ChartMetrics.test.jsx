/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */



import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import ChartMetrics from './ChartMetrics';

const items = [
  { id: 'checkbox-1', name: 'Metric 1', checked: false },
  { id: 'checkbox-2', name: 'Metric 2', checked: true },
  { id: 'checkbox-3', name: 'Metric 3', checked: false },
];

const props = {
  items,
  onChange: jest.fn(),
  limit: 2,
  placeholder: 'Select metrics',
};

describe('ChartMetrics', () => {
  test('Renders correctly', () => {
    const { container } = render(<ChartMetrics {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('Uses correct metrics name', () => {
    const { getByText } = render(<ChartMetrics {...props} />);

    userEvent.click(getByText('Select metrics'));

    expect(getByText('Metric 1')).not.toBeFalsy();
    expect(getByText('Metric 2')).not.toBeFalsy();
    expect(getByText('Metric 3')).not.toBeFalsy();
  });

  test('Sets the width if prop is passed', () => {
    const { getByTestId } = render(<ChartMetrics {...props} width={600} />);
    expect(getByTestId('chart-metrics')).toHaveStyle('maxWidth: 600px');
  });

  test('Toggles values correctly when limit 1', () => {
    const { getByText } = render(<ChartMetrics {...props} limit={1} />);

    userEvent.click(getByText('Metric 2'));

    userEvent.click(getByText('Metric 1'));

    expect(props.onChange).toHaveBeenCalledTimes(1);

    expect(props.onChange).toHaveBeenCalledWith({ id: 'checkbox-1', name: 'Metric 1' });

    props.onChange.mockReset();
  });

  test('Toggles values correctly', () => {
    const { getByText } = render(<ChartMetrics {...props} />);

    userEvent.click(getByText('Select metrics'));

    userEvent.click(getByText('Metric 1'));
    userEvent.click(getByText('Metric 2'));
    userEvent.click(getByText('Metric 3'));

    expect(props.onChange).toHaveBeenCalledTimes(3);

    const {
      mock: { calls },
    } = props.onChange;
    const lastCall = calls.pop();

    expect(lastCall).toEqual([
      [
        { id: 'checkbox-1', name: 'Metric 1' },
        { id: 'checkbox-3', name: 'Metric 3' },
      ],
    ]);
  });
});
