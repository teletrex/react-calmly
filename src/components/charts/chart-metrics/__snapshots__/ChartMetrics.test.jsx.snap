// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartMetrics Renders correctly 1`] = `
<div
  class="bx--chart-metrics bx--chart-metrics-multiple"
  data-testid="chart-metrics"
>
  <div
    class="bx-proposal-dropdown-wrapper bx-proposal-dropdown-wrapper--multiselect"
  >
    <div
      class="bx-proposal-dropdown__label"
    >
      <span>
        Metrics
      </span>
    </div>
    <div
      class="bx-proposal-dropdown__helper-text"
    >
      <span>
        Select up to 2 metrics
      </span>
    </div>
    <div
      class="bx-proposal-dropdown"
      role="listbox"
    >
      <span>
        <div
          aria-label="Select metrics"
          class="bx-proposal-dropdown__field bx-proposal-dropdown__field--multiselect"
          role="button"
          tabindex="0"
        >
          <span
            class="bx-proposal-dropdown__field__label"
          >
            <div
              class="bx--tag bx--tag--filter bx--tag--filter bx--tag--oval"
              title=""
            >
              <div
                class="bx--tag__label"
              >
                <span
                  class="bx--tag__text-ellipsis"
                >
                  1
                </span>
              </div>
              <button
                aria-label="Remove"
                title="Remove"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  focusable="false"
                  height="16"
                  preserveAspectRatio="xMidYMid meet"
                  style="will-change: transform;"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 4.7l-.7-.7L8 7.3 4.7 4l-.7.7L7.3 8 4 11.3l.7.7L8 8.7l3.3 3.3.7-.7L8.7 8z"
                  />
                </svg>
              </button>
            </div>
            Select metrics
          </span>
          <div
            class="bx-proposal-dropdown__field__menu-icon"
          >
            <svg
              aria-hidden="true"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              style="will-change: transform;"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 11L3 6l.7-.7L8 9.6l4.3-4.3.7.7z"
              />
            </svg>
          </div>
        </div>
      </span>
    </div>
  </div>
</div>
`;
